charon/src/ast/types.rs

```diff
@@ -333,20 +333,41 @@ pub struct VariantLayout {
     /// The offset of each field.
     #[drive(skip)]
     pub field_offsets: Vector<FieldId, ByteCount>,
+    /// Whether this variant is uninhabited. If `field_offsets` is empty
+    /// this could also mean that the variant is a ZST. Thus we need this flag
+    /// additionally.
+    #[drive(skip)]
+    pub uninhabited: bool,
+    /// The memory representation of the discriminant corresponding to this
+    /// variant. It must be of the same type as the corresponding [`DiscriminantLayout::tag_ty`].
+    ///
+    /// If it's `None`, then this variant is either the untagged variant (cf. [`TagEncoding::Niche::untagged_variant`])
+    /// containing the relevant niche or the single variant of a struct or uninhabited.
+    #[drive(skip)]
+    pub tag: Option<ScalarValue>,
 }
 
-/// Layout of the discriminant with its offset and representation type.
-///
-/// Does not include information about the value range.
+/// Describes how the discriminant is related to the tag, i.e. its memory representation.
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-pub enum DiscriminantLayout {
-    Direct {
-        /// The offset of the discriminant in bytes.
-        offset: ByteCount,
-        /// The representation type of the discriminant.
-        repr: IntegerTy,
-    },
-    Niche, // TODO: Add more useful information about niches here in the future.
+pub enum TagEncoding {
+    /// Represents the direct encoding of the discriminant as the tag via integer casts.
+    Direct,
+    /// Represents the encoding of the discriminant in the niche of variant `untagged_variant`.
+    Niche { untagged_variant: VariantId },
+}
```

Comment:
That feels redundant with the `tag` field of each variant. The only sensible way to read/write a discriminant needs to go through that `tag` field so I think this enum can be removed.